src/plugins/kibana/server/routes/api/ingest/register_bulk.js

```diff
@@ -0,0 +1,98 @@
+import { Promise } from 'bluebird';
+import { parse, transform } from 'csv';
+import _ from 'lodash';
+import hi from 'highland';
+import { patternToIngest } from '../../../../common/lib/convert_pattern_and_ingest_name';
+import { PassThrough } from 'stream';
+import bulkRequestSchema from '../../../lib/schemas/bulk_request_schema';
+
+export function registerBulk(server) {
+  server.route({
+    path: '/api/kibana/{id}/_bulk',
+    method: 'POST',
+    config: {
+      payload: {
+        output: 'stream',
+        maxBytes: 1024 * 1024 * 1024
+      },
+      validate: {
+        payload: bulkRequestSchema
+      }
+    },
+    handler: function (req, reply) {
+      const boundCallWithRequest = _.partial(server.plugins.elasticsearch.callWithRequest, req);
+      const indexPattern = req.params.id;
+      const usePipeline = req.payload.pipeline;
+      const csv = req.payload.csv;
+      const fileName = csv.hapi.filename;
+      const responseStream = new PassThrough();
+      const parser = parse({
+        columns: true,
+        auto_parse: true,
+        delimiter: _.get(req.payload, 'delimiter', ',')
+      });
+
+      let currentLine = 2; // Starts at 2 since we parse the header separately
+
+      responseStream.write('[');
+      csv.pipe(parser);
+
+      hi(parser)
+      .consume((err, doc, push, next) => {
+        if (err) {
+          push(err, null);
+          next();
+        }
+        else if (doc === hi.nil) {
+          // pass nil (end event) along the stream
+          push(null, doc);
+        }
+        else {
+          push(null, {index: {_id: `L${currentLine} - ${fileName}`}});
+          push(null, doc);
+          currentLine++;
+          next();
+        }
+      })
+      .batch(2000)
```

Comment:
I tested it before and such small batches seemed to be significantly slower, but I just re-tested and I didn't see the same slowness so maybe it was just a hiccup before. I lowered the bulk size to 100, but you'll see `.batch(200)` in the code because that includes the documents + the bulk action json.